File: 백준/Gold/2143. 두 배열의 합/두 배열의 합.cc
#include <iostream>
#include <algorithm>
#include <map>
using namespace std;

/* define */
#define endl '\n'
#define ll long long

int N, T, M;
int arr[1000+1];
int brr[1000 + 1];
int apsum[1000 + 1];
int bpsum[1000 + 1];
map<ll, ll> A, B; // {합, 갯수}
void init() {
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cin >> T;
	cin >> N;
	for (int i = 0; i < N; i++) cin >> arr[i];
	cin >> M;
	for (int i = 0; i < M; i++) cin >> brr[i];

	// get prefix
	for (int i = 1; i <= N; i++) {
		apsum[i] = apsum[i - 1] + arr[i - 1];
	}
	for (int i = 1; i <= M; i++) {
		bpsum[i] = bpsum[i - 1] + brr[i - 1];
	}

	// 부분배열합
	
	for (int i = 1; i <= N; i++) {
		for(int j=1; j<=i; j++){
			A[apsum[i] - apsum[j - 1]]++;
		}
	}
	for (int i = 1; i <= M; i++) {
		for (int j = 1; j <= i; j++) {
			B[bpsum[i] - bpsum[j - 1]]++;
		}
	}
	
	// A를 기점으로,  B를 이분탐색
	ll ans = 0;
	for (auto iter = A.begin(); iter != A.end(); iter++) {
		auto res = B.lower_bound((ll)T - iter->first);
		if (res != B.end() && (res->first + iter->first) == T) {
			ans += res->second * iter->second;
		}

	}

	cout << ans << endl;
}
int main() {
	init();

	return 0;
}

File: 백준/Gold/2143. 두 배열의 합/두 배열의 합.java
import java.util.*;
import java.io.*;

public class Main {
    static BufferedReader br;
    static StringTokenizer st;
    static StringBuilder sb;
    static int T, N, M;
    static int arr[], brr[];
    static int[] apsum, bpsum;
    public static void main(String[] args) throws Exception{
        // init
        br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine(), " ");
        T = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(br.readLine());
        arr = new int[N];
        st = new StringTokenizer(br.readLine(), " ");
        for(int i=0;i<N;i++) arr[i] = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(br.readLine());
        brr = new int[M];
        st = new StringTokenizer(br.readLine(), " ");
        for(int i=0;i<M;i++) brr[i] = Integer.parseInt(st.nextToken());

        // prefix
        apsum = new int[N+1];
        bpsum = new int[M+1];

        for(int i=1;i<=N;i++) apsum[i] = apsum[i-1] + arr[i-1];
        for(int i=1;i<=M;i++) bpsum[i] = bpsum[i-1] + brr[i-1];

        // get subset of sum
        HashMap<Integer, Integer> amp = new HashMap<>();
        HashMap<Integer, Integer> bmp = new HashMap<>();
        for(int i=0;i<=N;i++){
            for(int j=i+1; j<=N;j++){
                amp.put(apsum[j]- apsum[i], amp.getOrDefault(apsum[j] - apsum[i], 0) + 1);
            }
        }
        for(int i=0;i<=M;i++){
            for(int j=i+1;j<=M;j++){
                bmp.put(bpsum[j] - bpsum[i], bmp.getOrDefault(bpsum[j] - bpsum[i], 0) + 1);
            }
        }


        // search
        long answer =0;
        for(Integer key : amp.keySet()){
            if(!bmp.containsKey(T - key)) continue;
            answer += (long) amp.get(key) * bmp.get(T-key);
        }
        System.out.println(answer);
    }
}